File: dotnet/rapyd-sample/Controllers/ApiController.cs
using Microsoft.AspNetCore.Mvc;
using rapyd_sample.models;
using System.Text;
using System.Text.Json;

namespace rapyd_sample.Controllers
{
    [ApiController]
    [Route("/api")]
    public class ApiController : ControllerBase
    {
        private readonly ILogger<ApiController> _logger;
        private readonly RapydClient _client;
        private readonly WebhookEventsService _webhookEventsService;

        public ApiController(ILogger<ApiController> logger, RapydClient rapydClient, WebhookEventsService webhookEventsService)
        {
            _logger = logger;
            _client = rapydClient;
            _webhookEventsService = webhookEventsService;
        }

        [HttpGet("customers")]
        public async Task<ActionResult<IList<Customer>>> GetCustomers()
        {
            var customers = await _client.GetCustomers();
            return Ok(customers);
        }

        [HttpPost("customers")]
        public async Task<ActionResult<Customer>> CreateCustomer([FromBody] CreateCustomerBody body)
        {
            var customer = await _client.CreateCustomer(body);
            return Ok(customer);
        }

        [HttpGet("customers/{customerId}/paymentMethods")]
        public async Task<ActionResult<List<CustomerPaymentMethod>>> GetCustomerPaymentMethods([FromRoute] string customerId)
        {
            var customer = await _client.GetCustomersPaymentMethods(customerId);
            return Ok(customer);
        }

        [HttpGet("paymentMethods")]
        public async Task<ActionResult<IList<PaymentMethod>>> GetPaymentMethods([FromQuery] string country)
        {
            var methods = await _client.GetPaymentMethods(country);
            return Ok(methods);
        }

        [HttpGet("paymentMethodRequiredFields/{type}")]
        public async Task<ActionResult<PaymentMethodRequiredFields>> GetPaymentMethodRequiredFields([FromRoute] string type)
        {
            var methods = await _client.GetPaymentMethodRequiredFields(type);
            return Ok(methods);
        }

        [HttpPost("checkout")]
        public async Task<ActionResult<Checkout>> CreateCheckout([FromBody] CreateCheckoutBody body)
        {
            var checkout = await _client.CreatePaymentCheckout(body);
            return Ok(checkout);
        }

        [HttpPost("payment")]
        public async Task<ActionResult<Payment>> CreatePayment([FromBody] CreatePaymentBody body)
        {
            var payment = await _client.CreatePayment(body);
            return Ok(payment);
        }

        [HttpDelete("payment/{paymentId}")]
        public async Task<ActionResult<Payment>> CancelPayment([FromRoute] string paymentId)
        {
            var payment = await _client.CancelPayment(paymentId);
            return Ok(payment);
        }

        [HttpPost("webhook")]
        public async Task<IActionResult> Webhook([FromHeader] long timestamp, [FromHeader] string salt, [FromHeader] string signature)
        {
            var body = string.Empty;

            using (StreamReader reader = new StreamReader(Request.Body, Encoding.UTF8))
            {
                body = await reader.ReadToEndAsync();
            }

            var url = $"http://{Request.Host.Host}{Request.Path}";
            if (!_client.AuthWebhookRequest(signature, url, salt, timestamp, body)) return Unauthorized();

            try
            {
                JsonDocument document = JsonDocument.Parse(body);
                var id = document.RootElement.GetProperty("id").GetString();
                var type = document.RootElement.GetProperty("type").GetString();
                var data = document.RootElement.GetProperty("data").Deserialize<object>();
                _webhookEventsService.tryAddEvent(id, type, data);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error while handling webhook event");
            }
            return Ok();
        }

        [HttpGet("webhook/events")]
        public ActionResult<IEnumerable<WebhookEvent>> GetWebhookEvents()
        {
            return Ok(_webhookEventsService.getAllEvents());
        }
    }
}

File: nodejs/src/controllers/apiController.ts
import {
  Param,
  Body,
  Get,
  Post,
  Delete,
  HttpCode,
  UseBefore,
  JsonController,
  QueryParam,
  Req,
  HeaderParam,
  Res,
  OnUndefined,
} from 'routing-controllers';
import { OpenAPI, ResponseSchema } from 'routing-controllers-openapi';

import { validationMiddleware } from '@middlewares/validation.middleware';
import RapydService from '@/services/rapydService';
import { CreateCustomerBody, Customer } from '@/models/customers';
import { CreateCheckoutBody } from '@/models/checkout';
import { CreatePaymentBody } from '@/models/payment';
import webhookEventService from '@/services/webhookEventService';
import { Request, Response } from 'express';
import { HttpException } from '@/exceptions/HttpException';
import { WebhookEvent } from '@/models/webhookEvent';
import { PaymentMethod } from '@/models/paymentMethod';
import { PaymentMethodRequiredFields } from '@/models/paymentMethodRequiredFields';

@JsonController('/api')
export class ApiController {
  @Get('/customers')
  @OpenAPI({ summary: 'Return a list of customers' })
  @ResponseSchema(Customer, { isArray: true })
  async getCustomers() {
    const rapydService = new RapydService();
    const customers = await rapydService.getCustomers();
    return customers;
  }

  @Post('/customers')
  @HttpCode(200)
  @UseBefore(validationMiddleware(CreateCustomerBody, 'body'))
  @ResponseSchema(Customer)
  @OpenAPI({ summary: 'Create a new customer' })
  async createCustomer(@Body() body: CreateCustomerBody) {
    const rapydService = new RapydService();
    const customer = await rapydService.createCustomer(body);
    return customer;
  }

  @Get('/customers/:customerId/paymentMethods')
  @OpenAPI({ summary: 'Return a list of customer payment methods' })
  @ResponseSchema(PaymentMethod, { isArray: true })
  async getCustomerPaymentMethods(@Param('customerId') customerId: string) {
    const rapydService = new RapydService();
    const customerPaymentMethods = await rapydService.getCustomerPaymentMethods(customerId);
    return customerPaymentMethods;
  }

  @Get('/paymentMethods')
  @OpenAPI({ summary: 'Return a list of payment methods' })
  @ResponseSchema(PaymentMethod, { isArray: true })
  async getPaymentMethods(@QueryParam('country') country: string) {
    const rapydService = new RapydService();
    const paymentMethods = await rapydService.getPaymentMethods(country);
    return paymentMethods;
  }

  @Get('/paymentMethodRequiredFields/:type')
  @OpenAPI({ summary: 'Return a list of required field for given payment method' })
  @ResponseSchema(PaymentMethodRequiredFields)
  async getPaymentMethodRequiredFields(@Param('type') type: string) {
    const rapydService = new RapydService();
    const paymentRequiredFields = await rapydService.getPaymentMethodsRequiredFields(type);
    return paymentRequiredFields;
  }

  @Post('/checkout/redirect')
  @OnUndefined(303)
  @UseBefore(validationMiddleware(CreateCheckoutBody, 'body'))
  @OpenAPI({ summary: 'redirects to a new checkout page' })
  async redirectToCheckout(@Res() res: Response, @Body() body: CreateCheckoutBody) {
    const rapydService = new RapydService();
    const redirectUrl = await rapydService.createCheckoutRedirectURL(body);
    res.redirect(303, redirectUrl);
  }

  @Post('/checkout')
  @HttpCode(200)
  @UseBefore(validationMiddleware(CreateCheckoutBody, 'body'))
  @OpenAPI({ summary: 'Create a new payment checkout' })
  async createCheckout(@Body() body: CreateCheckoutBody) {
    const rapydService = new RapydService();
    return await rapydService.createPaymentCheckout(body);
  }

  @Post('/payment')
  @HttpCode(200)
  @UseBefore(validationMiddleware(CreatePaymentBody, 'body'))
  @OpenAPI({ summary: 'Create a payment' })
  async createPayment(@Body() body: CreatePaymentBody) {
    const rapydService = new RapydService();
    const payment = await rapydService.createPayment(body);
    return payment;
  }

  @Delete('/payment/:paymentId')
  @HttpCode(200)
  @OpenAPI({ summary: 'Cancel a payment' })
  async cancelPayment(@Param('paymentId') paymentId: string) {
    const rapydService = new RapydService();
    const payment = await rapydService.cancelPayment(paymentId);
    return payment;
  }

  @Get('/webhook/events')
  @OpenAPI({ summary: 'Return a list of events' })
  @ResponseSchema(WebhookEvent, { isArray: true })
  async getEvents() {
    const events = webhookEventService.getAllEvents();

    return events;
  }

  @Post('/webhook')
  @OpenAPI({ summary: 'Saves triggered rapyd actions' })
  async saveEvent(
    @Req() request: Request,
    @HeaderParam('signature') signature: string,
    @HeaderParam('timestamp') timestamp: number,
    @HeaderParam('salt') salt: string,
    @Body() body: any,
  ) {
    const rapydService = new RapydService();
    const url = `${request.protocol}://${request.hostname}${request.path}`;
    if (!rapydService.authWebhookRequest(signature, url, salt, timestamp, body)) {
      throw new HttpException(401, 'Signature not valid');
    }
    const eventId = body.id;
    const eventType = body.type;
    const data = body.data;

    if (!eventId || !eventType || !data) {
      throw new HttpException(400, 'Bad Request');
    }
    const status = webhookEventService.tryAddEvent(eventId, eventType, data);

    return status;
  }
}
